maintainers/link-headers

```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python3
+
+# This script must be run from the root of the Nix repository.
+#
+# For include path hygiene, we need to put headers in a separate
+# directory than sources. But during development, it is nice to paths
+# that are similar for headers and source files, e.g.
+# `foo/bar/baz.{cc,hh}`, e.g. for less typing when opening one file, and
+# then opening the other file.
+#
+# This script symlinks the headers next to the source files to
+# facilitate such a development workflows. It also updates
+# `src/.gitignore` so that the symlinks are not accidentally committed
+# by mistake.
+
+import os
+from pathlib import Path
+import subprocess
+
+# Path to the source directory
+SRC_DIR = Path("src")
+
+# Get header files from git
+result = subprocess.run(
+    ["git", "-C", str(SRC_DIR), "ls-files", "*/include/nix/**.hh"],
+    text=True, capture_output=True, check=True
+)
+header_files = result.stdout.strip().split("\n")
+header_files.sort()
+
+# Build header pairs
+header_pairs = []
+for file in header_files:
+    project, header = file.split("/include/nix/", 1)
+    header_pairs.append((Path(project), Path(header)))
+
+# Generate .gitignore file
+gitignore_path = SRC_DIR / ".gitignore"
+with open(gitignore_path, "w") as gitignore:
+    gitignore.write("# DO NOT COMMIT! Autogenerated\n")
+    gitignore.write("# Symlinks for headers to be next to sources for development\n")
+    gitignore.write('# Run "maintainers/link-headers" to regenerate\n\n')
+
+    for project, header in header_pairs:
+        gitignore.write(f"/{project / header}\n")
+
+for project, header in header_pairs:
+    # Reconstruct the full path (relative to SRC_DIR) to the header file.
+    file = project / "include" / "nix" / header
+
+    # The symlink should be created at "project/header", i.e. next to the project's sources.
+    link = project / header
+
+    # Compute a relative path from the symlink's parent directory to the actual header file.
+    relative_source = os.path.relpath(
+        SRC_DIR / file,
+        SRC_DIR / link.parent
+    )
+
+    # Create the symbolic link.
+    full_link_path = SRC_DIR / link
+    os.makedirs(full_link_path.parent, exist_ok=True)
+    if os.path.islink(full_link_path):
+        os.remove(full_link_path)
+    os.symlink(relative_source, full_link_path)
```

Comment:
Btw this would break on Windows btw. Making a symlink requires root.

Reply:
This script is for developer convenience and completely optional.
I personally don't expect to make use of it.